Automattic/woocommerce-payments
# Streamline checkout classes

Closes https://github.com/Automattic/woocommerce-payments/issues/7783
Contributes to https://github.com/Automattic/woocommerce-payments/issues/6882 as per first bullet point [here](https://github.com/Automattic/woocommerce-payments/issues/6882#issuecomment-1789258356)

#### Changes proposed in this Pull Request
This PR merges two checkout classes into one
```
WC_Payments_Checkout & WC_Payments_UPE_Checkout extends WC_Payments_Checkout --> WC_Payments_Checkout
```
Since the latest UPE rollout, only `WC_Payments_UPE_Checkout` was used. It also used parent class methods in some places.

As a result, in this PR
* all the logic from `WC_Payments_Checkout` which was used in `WC_Payments_UPE_Checkout` was moved into `WC_Payments_UPE_Checkout`
* every unused method from `WC_Payments_Checkout` was removed
* `WC_Payments_Checkout` was removed, as the previous two points led to this class being empty
* `WC_Payments_UPE_Checkout` was renamed to `WC_Payments_Checkout` to reduce complexity and improve readability
> 💡 GitHub shows changes in `WC_Payments_UPE_Checkout` as if they were applied to the `WC_Payments_Checkout`, because the former was renamed to the latter and GitHub has hard time displaying the changes in the right way. To make the review process easier, you can review [this commits range](https://github.com/Automattic/woocommerce-payments/pull/7869/files/ce1de59fec3b58b38ecd86fe4cf78ae504ba0dbc) first and [then review the name change separately](https://github.com/Automattic/woocommerce-payments/pull/7869/commits/3ea86093b97dba7755e68d8698ea92cde5f74b13).

Another significant step forward was the unit tests refactoring and extension. Multiple dependencies to the checkout class within the gateway test suite made the tests mock their own units under test. Checkout classes were used in unit tests for gateways, where we directly or indirectly tested their implementation from the gateway level, which is wrong. This PR improves this and to confirm this, you might try to search for `WC_Payments_Checkout` in test files and find that it's being used only in the test suite which is supposed to test checkout. Also a few new test cases were added to `test-class-wc-payments-checkout.php`.









#### Changes and testing instructions
* [remove deprecated hooks and take care of zero order total scripts reg](https://github.com/Automattic/woocommerce-payments/pull/7869/commits/26a873f68b5a809d42d4894008506dd68fa0e9ab)
    * test zero order total with free shipping & change the country to non-free shipping country without leaving the 
        *  checkout page and confirm payment methods are loaded
        *  ensure you have two shipping zones in WooCommerce (e.g. one free shipping for US and another one flat rate $10 for e.g. United Arab Emirates)
        *  add one free physical product and open the shortcode checkout so that US is chosen by default as shipping address country
        *  confirm no payment methods section is loaded
        *  change the shipping country to United Arab Emirates and confirm that payment methods are loading properly
* [streamline get_payment_fields_js_config into single implementation](https://github.com/Automattic/woocommerce-payments/pull/7869/commits/c23916318ececc6fdab85bbe92d6b4971c309dd5) & 5e5889aa9366098013eeb32370be4c7ae7f6d3a7 & 33200e61dfc807eb439814665bab9fc2877b9d62
    * open shortcode checkout with multiple PMs enabled, open browser console & confirm no js error are shown and every PM is displayed, perform two purchases: one with cc, another with non-cc
    * repeat the above with Blocks checkout
* disable every PM other than CC and pay with card on Blocks & shortcode
* confirm `save payment method` checkbox is displayed for logged in user for card PM and is not displayed for other, non reusable PMs

-------------------


**Post merge**
